File: ThalesMobileSimulator/ThalesCoreSource/HostCommands/CommandExplorer.vb
''
'' This program is free software; you can redistribute it and/or modify
'' it under the terms of the GNU General Public License as published by
'' the Free Software Foundation; either version 2 of the License, or
'' (at your option) any later version.
''
'' This program is distributed in the hope that it will be useful,
'' but WITHOUT ANY WARRANTY; without even the implied warranty of
'' MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'' GNU General Public License for more details.
''
'' You should have received a copy of the GNU General Public License
'' along with this program; if not, write to the Free Software
'' Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
'' 

Imports System.Reflection

Namespace HostCommands

    ''' <summary>
    ''' Used to find class that implement host commands.
    ''' </summary>
    ''' <remarks>
    ''' The command explorer is used to find classes amongst the loaded assemblies that
    ''' have the <see cref="HostCommands.ThalesCommandCode"/> attribute.
    ''' </remarks>
    Public Class CommandExplorer

        Private _commandTypes As New SortedList(Of String, CommandClass)

        ''' <summary>
        ''' CommandExplorer constructor.
        ''' </summary>
        ''' <remarks>
        ''' The constructor will search the loaded assemblies for classes that have the
        ''' <see cref="HostCommands.ThalesCommandCode"/> attribute.
        ''' </remarks>
        Public Sub New()

            ''We assume a single assembly for the mobile version.
            Dim asm() As [Assembly] = New [Assembly]() {System.Reflection.Assembly.GetExecutingAssembly()}
            For i As Integer = 0 To asm.GetUpperBound(0)
                Dim t As Type
                For Each t In asm(i).GetTypes()
                    Dim atr As Attribute
                    For Each atr In t.GetCustomAttributes(False)
                        If atr.GetType() Is GetType(ThalesSim.Core.HostCommands.ThalesCommandCode) Then
                            Try
                                _commandTypes.Add(CType(atr, ThalesCommandCode).CommandCode, _
                                                  New CommandClass(CType(atr, ThalesCommandCode).CommandCode, _
                                                                   CType(atr, ThalesCommandCode).ResponseCode, _
                                                                   CType(atr, ThalesCommandCode).ResponseCodeAfterIO, _
                                                                   t, asm(i).FullName, _
                                                                   CType(atr, ThalesCommandCode).Description))
                            Catch ex As ArgumentException
                                'We ignore attempts to add duplicates - this may happen when running the unit tests.
                            End Try
                        End If
                        'End If
                    Next
                Next
            Next

        End Sub

        ''' <summary>
        ''' Returns a summary of all loaded host commands.
        ''' </summary>
        ''' <remarks>
        ''' The method returns a summary description of all loaded host commands (classes that
        ''' inherit from <see cref="HostCommands.AHostCommand"/> and declare the 
        ''' <see cref="HostCommands.ThalesCommandCode"/> attribute).
        ''' </remarks>
        Public Function GetLoadedCommands() As String
            Dim s As String = ""
            Dim en As IEnumerator(Of KeyValuePair(Of String, CommandClass)) = _commandTypes.GetEnumerator()

            'Removed unnecessary Reset() for Mono compatibility.
            'en.Reset()

            While en.MoveNext
                s = s + "Command code: " + en.Current.Value.CommandCode + vbCrLf + _
                        "Response code: " + en.Current.Value.ResponseCode + vbCrLf + _
                        "Type: " + en.Current.Value.DeclaringType.FullName() + vbCrLf + _
                        "Description: " + en.Current.Value.Description + vbCrLf + vbCrLf
            End While
            en.Dispose()
            en = Nothing
            Return s
        End Function

        ''' <summary>
        ''' Returns a <see cref="CommandClass"/> object for a specified command.
        ''' </summary>
        ''' <remarks>
        ''' Returns a <see cref="CommandClass"/> object for a specified command. If the
        ''' command is not implemented, Nothing is returned.
        ''' </remarks>
        Public Function GetLoadedCommand(ByVal commandCode As String) As CommandClass
            Try
                Return _commandTypes(commandCode)
            Catch ex As KeyNotFoundException
                Return Nothing
            End Try
        End Function

        ''' <summary>
        ''' Clears the sorted list with the loaded commands.
        ''' </summary>
        ''' <remarks></remarks>
        Public Sub ClearLoadedCommands()
            _commandTypes.Clear()
        End Sub

    End Class

End Namespace

File: ThalesSim.Core/Commands/CommandExplorer.cs
/*
 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 2 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

using System;
using System.Collections.Generic;
using System.Text;
using ThalesSim.Core.Commands.Console;
using ThalesSim.Core.Commands.Host;

namespace ThalesSim.Core.Commands
{
    /// <summary>
    /// The command explorer is used to keep track of loaded
    /// types that implement host or console commands.
    /// </summary>
    public class CommandExplorer
    {
        private static readonly SortedList<CommandType, SortedList<string, Command>> Commands =
            new SortedList<CommandType, SortedList<string, Command>>();

        /// <summary>
        /// Discover host and console commands.
        /// </summary>
        public static void Discover()
        {
            var assemblies = AppDomain.CurrentDomain.GetAssemblies();
            foreach (var asm in assemblies)
            {
                foreach (var t in asm.GetTypes())
                {
                    foreach (var attr in t.GetCustomAttributes(false))
                    {
                        if (attr.GetType() == typeof(ThalesConsoleCommandAttribute))
                        {
                            var consoleAttr = (ThalesConsoleCommandAttribute) attr;
                            var consoleCommand = new Command
                                                     {
                                                         Code = consoleAttr.CommandCode,
                                                         DeclaringType = t,
                                                         Assembly = asm,
                                                         Description = consoleAttr.Description,
                                                         Type = CommandType.Console
                                                     };

                            var needsAuth = t.GetCustomAttributes(typeof(AuthorizedStateAttribute), false);
                            consoleCommand.RequiresAuthorizedState = (needsAuth.Length != 0);

                            if (!Commands.ContainsKey(CommandType.Console))
                            {
                                Commands.Add(CommandType.Console, new SortedList<string, Command>());
                            }

                            try
                            {
                                Commands[CommandType.Console].Add(consoleCommand.Code, consoleCommand);
                            }
                            catch (ArgumentException)
                            {
                                // Nothing, unit tests.
                            }
                        }

                        if (attr.GetType() == typeof(ThalesHostCommandAttribute))
                        {
                            var hostAttr = (ThalesHostCommandAttribute) attr;
                            var hostCommand = new HostCommand
                                                  {
                                                      Assembly = asm,
                                                      Code = hostAttr.CommandCode,
                                                      DeclaringType = t,
                                                      Description = hostAttr.Description,
                                                      ResponseCode = hostAttr.ResponseCode,
                                                      ResponseCodeAfterIo = hostAttr.ResponseCodeAfterIo,
                                                      Type = CommandType.Host
                                                  };

                            var needsAuth = t.GetCustomAttributes(typeof (AuthorizedStateAttribute), false);
                            hostCommand.RequiresAuthorizedState = (needsAuth.Length != 0);

                            if (!Commands.ContainsKey(CommandType.Host))
                            {
                                Commands.Add(CommandType.Host, new SortedList<string, Command>());
                            }

                            try
                            {
                                Commands[CommandType.Host].Add(hostCommand.Code, hostCommand);
                            }
                            catch (ArgumentException)
                            {
                                // Nothing, unit tests.
                            }
                            
                        }
                    }
                }
            }
        }

        /// <summary>
        /// Get descriptions of discovered commands.
        /// </summary>
        /// <param name="commandType">Type of commands to look for.</param>
        /// <returns>String with discovered commands.</returns>
        public static string GetLoadedCommandDescriptions (CommandType commandType)
        {
            var sb = new StringBuilder();
            foreach (var key in Commands[commandType].Keys)
            {
                sb.Append(Commands[commandType][key].ToString());
                sb.AppendLine();
            }

            return sb.ToString();
        }

        /// <summary>
        /// Get a command.
        /// </summary>
        /// <param name="commandType">Type of command to search for.</param>
        /// <param name="code">Command code.</param>
        /// <returns>Found command or null.</returns>
        public static Command GetCommand (CommandType commandType, string code)
        {
            try
            {
                return Commands[commandType][code];
            }
            catch (KeyNotFoundException)
            {
                return null;
            }
        }
    }
}
